examples/vision/forwardforward.py

```diff
@@ -0,0 +1,352 @@
+        self.num_epochs = num_epochs
+
+    def call(self, x):
+        x_norm = tf.norm(x, ord=2, axis=1, keepdims=True)
+        x_norm = x_norm + 1e-4
+        x_dir = x / x_norm
+        res = self.dense(x_dir)
+        return self.relu(res)
+
+    def forwardforward(self, x_pos, x_neg):
+        loss_list = []
+        for i in trange(self.num_epochs):
+            with tf.GradientTape() as tape:
+                g_pos = tf.math.reduce_mean(tf.math.pow(self.call(x_pos), 2), 1)
+                g_neg = tf.math.reduce_mean(tf.math.pow(self.call(x_neg), 2), 1)
+
+                loss = tf.math.log(
+                    1
+                    + tf.math.exp(
+                        tf.concat([-g_pos + self.threshold, g_neg - self.threshold], 0)
+                    )
+                )
+                mean_loss = tf.math.reduce_mean(loss)
+                loss_list.append(mean_loss.numpy())
+            gradients = tape.gradient(mean_loss, self.trainable_weights)
+            self.optimizer.apply_gradients(zip(gradients, self.dense.trainable_weights))
+        return (
+            tf.stop_gradient(self.call(x_pos)),
+            tf.stop_gradient(self.call(x_neg)),
+            mean_loss,
+            loss_list,
+        )
+
+
+"""
+## Define the `FFNetwork` Custom Model
+
+With our custom layer defined, we also need to override the `train_step` method and
+define a custom `tf.keras.models.Model` that works with our `FFDense` layer.
+
+For this algorithm, we must 'embed' the labels onto the original image. To do so, we
+exploit the structure of MNIST images where the top-left 10 pixels are always zeros. We
+use that as a label space in order to visually one-hot-encode the labels within the image
+itself. This action is performed by the `overlay_y_on_x` function.
+
+We break down the prediction function with a per-sample prediction function which is then
+called over the entire test set by the overriden `predict()` function. The prediction is
+performed here with the help of measuring the `excitation` of the neurons per layer for
+each image. This is then summed over all layers to calculate a network-wide 'goodness
+score'. The label with the highest 'goodness score' is then chosen as the sample
+prediction.
+
+The `train_step` function is overriden to act as the main controlling loop for running
+training on each layer as per the number of epochs per layer.
+"""
+
+
+class FFNetwork(keras.Model):
+    def __init__(self, dims, layer_optimizer=keras.optimizers.Adam(learning_rate=0.03)):
+        super().__init__()
+        self.layer_optimizer = layer_optimizer
+        self.mean_loss = keras.metrics.Mean()
+        self.flatten_layer = keras.layers.Flatten()
+        self.layer_list = [keras.Input(shape=(dims[0],))]
+        for d in range(len(dims) - 1):
+            self.layer_list += [FFDense(dims[d + 1], optimizer=self.layer_optimizer)]
+
+    @tf.function()
+    def overlay_y_on_x(self, X, y):
+        x_res = X.numpy()
```

Comment:
Please don't go to numpy for this. For assignment you can use one of the solutions outlined in this thread: https://github.com/tensorflow/community/pull/433

Reply:
Ahh, this looks good, will try to incorporate this

Reply:
Would the use of the XLA op as you have mentioned in the issue be fine?